### PR TITLE
memory refactor

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/src/memory/base.ts [hunk 1]
```diff
@@ -13,3 +20,37 @@ export abstract class BaseMemory {
...
+): string {
+  const string_messages: string[] = [];
+  for (const m of messages) {
+    let role: string;
+    if (m instanceof HumanChatMessage) {
+      role = human_prefix;
+    } else if (m instanceof AIChatMessage) {
+      role = ai_prefix;
+    } else if (m instanceof SystemChatMessage) {
+      role = "System";
```
Comment (nfcampos) [inline langchain/src/memory/base.ts:47]: Why isn't this one configurable?

File: langchain/src/memory/chat_memory.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  chatHistory: ChatMessageHistory;
+
+  constructor(chatHistory?: ChatMessageHistory) {
+    super();
+    this.chatHistory = chatHistory ?? new ChatMessageHistory();
+  }
+
+  async saveContext(
+    inputValues: InputValues,
+    OutputValues: Promise<OutputValues>
```
Comment (nfcampos) [inline langchain/src/memory/chat_memory.ts:39]: This is a fairly strange API, receiving a promise of the argument. Should we take this opportunity to fix it?

**Review by nfcampos — APPROVED** [review]
LGTM

File: langchain/src/memory/chat_memory.ts (hunk 2 above)
Comment (agola11) [reply langchain/src/memory/chat_memory.ts:39]: yeah this is weird for sure, but changing might be involved. probs best suited for another PR. Could create issue?

Comment (nfcampos) [reply langchain/src/memory/chat_memory.ts:39]: yea fair enough

**Pull request merged**